text/0035-emergency-restarts.md

```diff
@@ -0,0 +1,116 @@
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The design of the network has had support for specifying upgrade points in the chainspec for quite some time. Such upgrade points are essentialy a specification of changes that need to be applied to the chain at some block heights.
+
+Coincidentally, that's exactly what we require for emergency restarts, too. As outlined in [CEP 6](https://github.com/CasperLabs/ceps/blob/master/text/0006-equivocation-catastrophes.md), after a network crashes and we identify the last safe block, we might need to:
+- replay some deploys on top of it
+- directly modify the set of validators applicable from the next block onwards
+- perform other changes to the global state as required.
+
+All these potential changes influence the global state, which means that they can be expressed as a set of transforms to be applied at a chosen block height.
+
+Currently, the post-state of a block at height `h` is defined in the following way:
+- It is the post-state of the block `h-1`
+- with the transforms resulting from executing the deploys contained in block `h` applied
+- with the transforms resulting from executing the `step` function applied, if block `h` contains the `EraEnd` structure.
+
+We need to add one more point to that:
+- with the transforms specified by the upgrade point at height `h` applied (if one exists).
+
+The criteria for validity of a block at height `h` should also take this into account, i.e. a block at height `h` is only valid if its global state hash is the result of applying the above steps to the global state specified at height `h-1`.
+
+Note that this has some desired properties:
+- We can specify any change to the state at any height without breaking the continuity of the chain. The first block after the restart is just like any other regular block, which means that eg. chain synchronization after the restart will work just like before the restart.
+- If the last safe block before the restart was at height `h`, but there were also blocks `h+1...h+n` that were finalized, they will become invalid after an upgrade point is specified at height `h+1`. Since the upgrade point wasn't included in the chainspec before the restart, the global state of the old block `h+1` won't satisfy the validity condition after an upgrade is added and the network is restarted. This means that all the blocks finalized after `h` will be discarded after restarting with the new chainspec and the network will resume operation at height `h+1`.
```

Comment:
But we should also take into account the possibility where we want to continue from a block at height `h` that in the view of some of the validators wasn't even seen as finalized.

I guess this can be achieved just by using the trusted hash?

Reply:
> I guess this can be achieved just by using the trusted hash?

I think so, yes.